src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs

```diff
@@ -163,7 +225,88 @@ private async Task GetProjectMetadataAsync(
+                {
+                    projectModel.AddProjectInformation(
+                        ProblemType.Warning,
+                        string.Format(CultureInfo.CurrentCulture, Strings.Warning_AuditSourceWithoutData, source.Name)
+                    );
+                }
+            }
+
+            return vulnerabilityInfo;
+        }
+
+        private static async Task<bool> TryAddSourceVulnerabilityInfo(
+            PackageSource source,
+            ListPackageReportModel reportModel,
+            ILogger logger,
+            CancellationToken cancellationToken,
+            List<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>> vulnerabilityInfo)
+        {
+            var repository = Repository.Factory.GetCoreV3(source);
+            var vulnerabilityProvider = new VulnerabilityInfoResourceV3Provider();
+            var (isCreated, resource) = await vulnerabilityProvider.TryCreate(repository, cancellationToken);
+
+            if (!isCreated || resource is not VulnerabilityInfoResourceV3 vulnerabilityResource)
+            {
+                return false;
+            }
+
+            reportModel.SourcesUsed.Add(source);
+
+            var vulnerabilityInfoResult = await vulnerabilityResource.GetVulnerabilityInfoAsync(
+                new SourceCacheContext(),
+                logger,
+                cancellationToken
+            );
+
+            if (vulnerabilityInfoResult?.KnownVulnerabilities != null)
+            {
+                vulnerabilityInfo.AddRange(vulnerabilityInfoResult.KnownVulnerabilities);
+            }
+
+            return true;
+        }
+
+
+        private static IEnumerable<PackageVulnerabilityMetadata> GetPackageVulnerabilities(
+            IEnumerable<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>> vulnerabilities,
+            string id,
+            string version)
+        {
+            if (vulnerabilities == null)
+            {
+                return Enumerable.Empty<PackageVulnerabilityMetadata>();
+            }
+
+            var parsedVersion = new NuGetVersion(version);
+            foreach (var vulnFile in vulnerabilities)
+            {
+                if (vulnFile.TryGetValue(id, out IReadOnlyList<PackageVulnerabilityInfo> vulnPackages) && vulnPackages != null)
+                {
+                    return vulnPackages
+                        .Where(package => package.Versions.Satisfies(parsedVersion))
+                        .Select(v => JsonExtensions.FromJson<PackageVulnerabilityMetadata>($"{{ \"AdvisoryUrl\": \"{v.Url}\", \"Severity\": \"{(int)v.Severity}\" }}"))
+                        .ToList();
+                }
+            }
+
+            return Enumerable.Empty<PackageVulnerabilityMetadata>();
+        }
+
+        private static void WarnForHttpSources(ListPackageArgs listPackageArgs, ListPackageProjectModel projectModel, IReadOnlyList<PackageSource> auditSources)
```

Comment:
There's duplication in this method now that we're checking two lists, we should add a local function that is run in both `foreach` blocks.